code/iaas/logic/src/main/java/io/cattle/platform/process/externalevent/ExternalEventCreate.java

```diff
@@ -0,0 +1,209 @@
+import io.cattle.platform.process.common.util.ProcessUtils;
+import io.cattle.platform.util.type.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Named
+public class ExternalEventCreate extends AbstractDefaultProcessHandler {
+
+    public static final String FIELD_AGENT_ID = "agentId";
+
+    private static final Logger log = LoggerFactory.getLogger(ExternalEventCreate.class);
+
+    @Inject
+    AccountDao accountDao;
+    @Inject
+    LockManager lockManager;
+    @Inject
+    ResourceMonitor resourceMonitor;
+    @Inject
+    GenericResourceDao resourceDao;
+    @Inject
+    StoragePoolDao storagePoolDao;
+    @Inject
+    VolumeDao volumeDao;
+    @Inject
+    HostDao hostDao;
+    @Inject
+    JsonMapper jsonMapper;
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        objectManager.children(null, null);
+        ExternalEvent event = (ExternalEvent)state.getResource();
+
+        if (StringUtils.isEmpty(event.getExternalId())) {
+            log.warn("External event doesn't have an external id: {}", event);
+            return null;
+        }
+
+        if (ExternalEventConstants.KIND_VOLUME_EVENT.equals(event.getKind())) {
+            handleVolumeEvent(event, state, process);
+        } else if (ExternalEventConstants.KIND_STORAGE_POOL_EVENT.equals(event.getKind())) {
+            handleStoragePoolEvent(event, state, process);
+        } else {
+            new IllegalStateException();
+        }
+
+        return null;
+    }
+
+    protected void handleVolumeEvent(final ExternalEvent event, ProcessState state, ProcessInstance process) {
+        lockManager.lock(new ExternalEventLock(VOLUME_POOL_LOCK_NAME, event.getAccountId(), event.getExternalId()), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                String spExtId = DataAccessor.fieldString(event, FIELD_SP_EXT_ID);
+                StoragePool storagePool = storagePoolDao.findStoragePoolByExternalId(event.getAccountId(), spExtId);
+                if (storagePool == null) {
+                    log.warn("Unknown storage pool. Returning. External id: {}", spExtId);
+                    return;
+                }
+                Volume volume = volumeDao.findVolumeByExternalId(storagePool.getId(), event.getExternalId());
+                switch (event.getEventType()) {
+                case ExternalEventConstants.TYPE_VOLUME_CREATE:
+                    if (volume == null) {
+                        Map<String, Object> volumeData = CollectionUtils.toMap(DataUtils.getFields(event).get("volume"));
+                        if (volumeData == null || volumeData.isEmpty()) {
```

Comment:
I believe "toMap" action never returns null; only an empty map, so the first check can be dropped